File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/propulsion_system/esc.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ...setup_assistant import SetupAssistant
    from ...parameter_getters import ParamGetterWidget

from overrides import override
from PyQt5.QtWidgets import QVBoxLayout

from ...parameter_getters import ParamGetterWidget_SpinBox, ParamGetterWidget_ComboBox
from .base import BaseSelectedLinkSettingWidget


class EscWidget(BaseSelectedLinkSettingWidget):
    NAME = "ESC"

    SIGNAL_MODE_MAP = {
        "BLHeli Open Loop": "blheli_open_loop",
        "BHLeli Closed Loop (Low Range)": "blheli_closed_loop_low_range",
        "BHLeli Closed Loop (Middle Range)": "blheli_closed_loop_mid_range",
        "BHLeli Closed Loop (High Range)": "blheli_closed_loop_high_range",
    }

    def __init__(self, main: SetupAssistant, link_name: str) -> None:
        super().__init__(main, link_name)

        rows = QVBoxLayout()
        self.setLayout(rows)

        self._param_rows = QVBoxLayout()
        rows.addLayout(self._param_rows)

        max_current_description = (
            "Maximum current that the ESC (Electronic Speed Controller) can safely handle. "
            "Exceeding this maximum current may lead to overheating or damage to the ESC, "
            "and in the worst case, it could cause failure or fire."
        )
        self._max_current = ParamGetterWidget_SpinBox(
            "Maximum Current", max_current_description, minimum=1, default=20, suffix=" A"
        )
        self._param_rows.addWidget(self._max_current)

        signal_mode_description = ""  # TODO
        self._signal_mode = ParamGetterWidget_ComboBox(
            "Signal Mode", signal_mode_description, self.SIGNAL_MODE_MAP.keys()
        )
        self._param_rows.addWidget(self._signal_mode)

        rows.addStretch()

    @override
    def is_valid(self) -> bool:
        return True

    @override
    def copy_from(self, src: EscWidget) -> None:
        for i in range(self._param_rows.count()):
            param_des: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param_src: ParamGetterWidget = src._param_rows.itemAt(i).widget()
            param_des.set(param_src.get())

    @override
    def dump_settings(self) -> dict:
        res = dict()
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            res[param.name()] = param.get()
        return res

    @override
    def load_settings(self, data: dict) -> None:
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param.set(data[param.name()])

    def max_current(self) -> float:
        return self._max_current.get()

    def signal_mode(self) -> str:
        return self.SIGNAL_MODE_MAP[self._signal_mode.get()]

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/propulsion_system/esc.cpp
#include <tobas_yaml_tools/convert/qstring.hpp>
#include <tobas_drone_core/esc.hpp>

#include "tobas_setup_assistant/setting_tabs/propulsion_system/esc.hpp"

namespace gui
{
namespace setup_assistant
{
namespace propulsion_system
{
ESCWidget::ESCWidget()
  : signal_mode_map_{ { "BLHeli Open Loop", tobas::esc_mode_t::BLHELI_OPEN_LOOP },
                      { "BHLeli Closed Loop (Low Range)", tobas::esc_mode_t::BLHELI_CLOSED_LOOP_LOW_RANGE },
                      { "BHLeli Closed Loop (Middle Range)", tobas::esc_mode_t::BLHELI_CLOSED_LOOP_MID_RANGE },
                      { "BHLeli Closed Loop (High Range)", tobas::esc_mode_t::BLHELI_CLOSED_LOOP_HIGH_RANGE } }
{
  const auto rows = new QVBoxLayout();
  setLayout(rows);

  max_current_ = new ParamGetterWidget_SpinBox(
    "Maximum Current", "Maximum current that the ESC (Electronic Speed Controller) can safely handle. "
                       "Exceeding this maximum current may lead to overheating or damage to the ESC, "
                       "and in the worst case, it could cause failure or fire.");
  max_current_->setMinimum(1);
  max_current_->setValue(20);
  max_current_->setSuffix(" A");
  rows->addWidget(max_current_);

  signal_mode_ = new ParamGetterWidget_ComboBox("Signal Mode", "");
  for (const auto& [text, _] : signal_mode_map_)
    signal_mode_->addChoice(text);
  rows->addWidget(signal_mode_);

  rows->addStretch();
}

const char* ESCWidget::name() const
{
  return "ESC";
}

bool ESCWidget::isValid()
{
  return true;
}

void ESCWidget::copyFrom(const BaseSelectedLinkSettingWidget* src)
{
  const auto derived = qobject_cast<const ESCWidget*>(src);
  max_current_->setValue(derived->max_current_->getValue());
  signal_mode_->setValue(derived->signal_mode_->getValue());
}

YAML::Node ESCWidget::dump() const
{
  YAML::Node node(YAML::NodeType::Map);

  node[max_current_->name()] = max_current_->getValue();
  node[signal_mode_->name()] = signal_mode_->getValue();

  return node;
}

void ESCWidget::load(const YAML::Node& node)
{
  max_current_->setValue(node[max_current_->name()].as<int>());
  signal_mode_->setValue(node[signal_mode_->name()].as<QString>());
}

double ESCWidget::maxCurrent() const
{
  return max_current_->getValue();
}

tobas::esc_mode_t ESCWidget::signalMode() const
{
  return signal_mode_map_.at(signal_mode_->getValue());
}
}  // namespace propulsion_system
}  // namespace setup_assistant
}  // namespace gui
